stored_requests/ab_fetcher.go

```diff
@@ -0,0 +1,212 @@
+	if requestData, impData, errs = f.fetcher.FetchRequests(ctx, requestIDs, impIDs); len(errs) > 0 || len(requestIDs) != 1 || len(requestData) == 0 {
+		return
+	}
+	requestID := requestIDs[0]
+	storedRequest := requestData[requestID]
+
+	// Extract ab_config json if present
+	abConfigs, dataType, _, err := jsonparser.Get(storedRequest, "ext", openrtb_ext.PrebidExtKey, "storedrequest", "ab_config")
+	if dataType == jsonparser.NotExist {
+		return
+	}
+	if err != nil {
+		errs = append(errs, fmt.Errorf(`failed to read ext.prebid.storedrequest.ab_config from storedrequest "%s": %v`, requestID, err))
+		return
+	}
+	if dataType != jsonparser.Array {
+		errs = append(errs, fmt.Errorf(`bad format for ext.prebid.storedrequest.ab_config in storedrequest "%s": expecting a list`, requestID))
+		return
+	}
+
+	// Run random selection and pick one ABConfig
+	abConfig, err := runABSelection(abConfigs)
+	if err != nil {
+		errs = append(errs, fmt.Errorf(`error parsing ab_config rules from storedrequest "%s": %v`, requestID, err))
+		return
+	}
+	if abConfig == nil {
+		return
+	}
+
+	// Build lists of new IDs to fetch for replacements
+	newImpIDs := []string{}
+	for _, impID := range impIDs {
+		if newImpID, replace := abConfig.ImpIDs[impID]; replace {
+			newImpIDs = append(newImpIDs, newImpID)
+		}
+	}
+	newReqIDs := []string{}
+	if abConfig.RequestID != requestID && len(abConfig.RequestID) > 0 {
+		newReqIDs = []string{abConfig.RequestID}
+	}
+
+	// Fetch replacement stored requests and imps
+	newRequestData, newImpData, newErrs := f.fetcher.FetchRequests(ctx, newReqIDs, newImpIDs)
+	if len(newErrs) > 0 {
+		errs = append(errs, fmt.Errorf(`errors fetching replacement stored data for ab_config`))
+		errs = append(errs, newErrs...)
+		return // or not, if we want to power through and have a partial replace
+	}
+
+	// Replace the stored data for original IDs with the new replacement values.
+	// This can be a little confusing (because the original ids are still present)
+	// but avoids rewriting the main request json to replace the ids
+	var analyticsInfo json.RawMessage
+	if value, found := newRequestData[abConfig.RequestID]; found {
+		storedRequest = value
+		// patch loaded stored request with original ab_config info to preserve it,
+		// and add ab_code for analytics
+		analyticsInfo = json.RawMessage(
+			fmt.Sprintf(`{"ext":{"%s":{"storedrequest":{"ab_config":%s,"ab_code":"%s"}}}}`,
+				openrtb_ext.PrebidExtKey, abConfigs, abConfig.Code))
+	} else {
+		// patch original stored request with just ab_code, the ab_config is already there
+		analyticsInfo = json.RawMessage(
+			fmt.Sprintf(`{"ext":{"%s":{"storedrequest":{"ab_code":"%s"}}}}`,
+				openrtb_ext.PrebidExtKey, abConfig.Code))
+	}
+	if enrichedRequest, err := jsonpatch.MergePatch(storedRequest, analyticsInfo); err == nil {
+		requestData[requestID] = enrichedRequest
```

Comment:
How expensive do you expect the MergePatch to be in terms of performance, when A/B Tests are enabled?

Reply:
PBS uses this MergePatch to merges stored request, and then each stored imp, where many keys match; this won't be slower than one of those... yeah we should bench it.

I'd prefer the code to go directly in the analytics context, but staying with the confines of a fetcher this was the best I could do.